### PR TITLE
MySql.Data - natively supported by library

### LINKED ISSUE #2542: Rewrite MySql.Data instrumentation to bytecode instrumenation
MySql.Data v 8.0.33 introduced breaking changes.
It is not feasible to longer use sourcecode instrumentation without some other magic (like Harmony https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1157).

Issue detected by #2478

The goal is to rewrite it to bytecode instrumentation.

### PR DESCRIPTION
**Pull request by Kielek**
## Why

Fixes #2542
Supersedes: #2686
Traces are natively supported by MySql.Data 8.1.0+

## What

Dropped reference to OpenTelemetry.Instrumentation.MySqlData..
Drop bytecode instrumentation for couple versions of MySql.Data.
Support native instrumentation for MySql.Data 8.1.0+.

It is last good moment to drop support for old versions as we are still in pre-GA mode.

## Tests

CI + manual verification.

Spans from our testing assembly.

```log
  + "InstrumentationScopeName = connector-net, Span = { "traceId": "JgX/gm5iJPSL47GzVkeSUA==", "spanId": "rsi9qPIsxU4=", "parentSpanId": "fX8Z9KbMbbA=", "name": "SQL Statement", "kind": "SPAN_KIND_CLIENT", "startTimeUnixNano": "1692003568358528800", "endTimeUnixNano": "1692003568480060900", "attributes": [ { "key": "db.system", "value": { "stringValue": "mysql" } }, { "key": "db.name", "value": { "stringValue": "" } }, { "key": "db.user", "value": { "stringValue": "root" } }, { "key": "db.statement", "value": { "stringValue": "SELECT @@max_allowed_packet, @@character_set_client, \n        @@character_set_connection, @@license, @@sql_mode, @@lower_case_table_names, @@autocommit;" } }, { "key": "thread.id", "value": { "intValue": "9" } } ], "status": { "code": "STATUS_CODE_OK" } }"
  + "InstrumentationScopeName = connector-net, Span = { "traceId": "JgX/gm5iJPSL47GzVkeSUA==", "spanId": "T4PGcAPH0Do=", "parentSpanId": "fX8Z9KbMbbA=", "name": "SQL Statement", "kind": "SPAN_KIND_CLIENT", "startTimeUnixNano": "1692003568493827600", "endTimeUnixNano": "1692003568520438800", "attributes": [ { "key": "db.system", "value": { "stringValue": "mysql" } }, { "key": "db.name", "value": { "stringValue": "" } }, { "key": "db.user", "value": { "stringValue": "root" } }, { "key": "db.statement", "value": { "stringValue": "SELECT TIMEDIFF(NOW(), UTC_TIMESTAMP())" } }, { "key": "thread.id", "value": { "intValue": "9" } } ], "status": { "code": "STATUS_CODE_OK" } }"
  + "InstrumentationScopeName = connector-net, Span = { "traceId": "JgX/gm5iJPSL47GzVkeSUA==", "spanId": "Dlxz2gsT24U=", "parentSpanId": "fX8Z9KbMbbA=", "name": "SQL Statement", "kind": "SPAN_KIND_CLIENT", "startTimeUnixNano": "1692003568529934000", "endTimeUnixNano": "1692003568547432200", "attributes": [ { "key": "db.system", "value": { "stringValue": "mysql" } }, { "key": "db.name", "value": { "stringValue": "" } }, { "key": "db.user", "value": { "stringValue": "root" } }, { "key": "db.statement", "value": { "stringValue": "SHOW COLLATION" } }, { "key": "thread.id", "value": { "intValue": "12" } }, { "key": "thread.name", "value": { "stringValue": ".NET ThreadPool Worker" } } ], "status": { "code": "STATUS_CODE_OK" } }"
  + "InstrumentationScopeName = connector-net, Span = { "traceId": "JgX/gm5iJPSL47GzVkeSUA==", "spanId": "W7zR1mJCiZw=", "parentSpanId": "fX8Z9KbMbbA=", "name": "SQL Statement", "kind": "SPAN_KIND_CLIENT", "startTimeUnixNano": "1692003568554957700", "endTimeUnixNano": "1692003568561390500", "attributes": [ { "key": "db.system", "value": { "stringValue": "mysql" } }, { "key": "db.name", "value": { "stringValue": "" } }, { "key": "db.user", "value": { "stringValue": "root" } }, { "key": "db.statement", "value": { "stringValue": "SET NAMES utf8mb4" } }, { "key": "thread.id", "value": { "intValue": "9" } } ], "status": { "code": "STATUS_CODE_OK" } }"
  + "InstrumentationScopeName = connector-net, Span = { "traceId": "JgX/gm5iJPSL47GzVkeSUA==", "spanId": "tWZB0VQ8ubk=", "parentSpanId": "fX8Z9KbMbbA=", "name": "SQL Statement", "kind": "SPAN_KIND_CLIENT", "startTimeUnixNano": "1692003568565478100", "endTimeUnixNano": "1692003568570570900", "attributes": [ { "key": "db.system", "value": { "stringValue": "mysql" } }, { "key": "db.name", "value": { "stringValue": "" } }, { "key": "db.user", "value": { "stringValue": "root" } }, { "key": "db.statement", "value": { "stringValue": "SET character_set_results=NULL" } }, { "key": "thread.id", "value": { "intValue": "9" } } ], "status": { "code": "STATUS_CODE_OK" } }"
  + "InstrumentationScopeName = connector-net, Span = { "traceId": "85728DPuaIjfcE4vBM3DHA==", "spanId": "KC7eEihzLwE=", "name": "SQL Statement", "kind": "SPAN_KIND_CLIENT", "startTimeUnixNano": "1692003568586073800", "endTimeUnixNano": "1692003568592915700", "attributes": [ { "key": "db.system", "value": { "stringValue": "mysql" } }, { "key": "db.name", "value": { "stringValue": "" } }, { "key": "db.user", "value": { "stringValue": "root" } }, { "key": "db.statement", "value": { "stringValue": "SELECT 123;" } }, { "key": "thread.id", "value": { "intValue": "9" } } ], "status": { "code": "STATUS_CODE_OK" } }"
  + "InstrumentationScopeName = connector-net, Span = { "traceId": "JgX/gm5iJPSL47GzVkeSUA==", "spanId": "fX8Z9KbMbbA=", "name": "Connection (pooled)", "kind": "SPAN_KIND_CLIENT", "startTimeUnixNano": "1692003567919962400", "endTimeUnixNano": "1692003568597012900", "attributes": [ { "key": "net.transport", "value": { "stringValue": "Socket" } }, { "key": "db.connection_string", "value": { "stringValue": "server=127.0.0.1;port=62329;user id=root" } }, { "key": "net.peer.port", "value": { "intValue": "62329" } } ], "status": { "code": "STATUS_CODE_OK" } }"
```

## Checklist

- [x] `CHANGELOG.md` is updated.
- [x] Documentation is updated.
- [x] New features are covered by tests.

### CODE REVIEW

**Review by nrcventura — APPROVED** [review]
This looks good, and is better than relying on bytecode instrumentation that we need to keep updated (especially as the semantic conventions stabilize). With the native instrumentation, do we think that .net framework will be available too?

File: src/CommonExcludedAssets.props [hunk 1]
```diff
@@ -7,17 +7,15 @@
...
-    <!-- MySql.Data is required by OpenTelemetry.Instrumentation.MySqlData. -->
-    <PackageReference Include="MySql.Data" ExcludeAssets="all" />
-    <!-- System.Configuration.ConfigurationManager is tranistive dependency required by OpenTelemetry.Instrumentation.MySqlData. -->
+    <!-- System.Configuration.ConfigurationManager is tranistive dependency required by OpenTelemetry.Instrumentation.StackExchangeRedis. -->
     <PackageReference Include="System.Configuration.ConfigurationManager" ExcludeAssets="all" />
-    <!-- System.Security.Cryptography.ProtectedData is tranistive dependency required by OpenTelemetry.Instrumentation.MySqlData. -->
+    <!-- System.Security.Cryptography.ProtectedData is tranistive dependency required by OpenTelemetry.Instrumentation.StackExchangeRedis. -->
     <PackageReference Include="System.Security.Cryptography.ProtectedData" ExcludeAssets="all" />
-    <!-- System.Drawing.Common is tranistive dependency required by OpenTelemetry.Instrumentation.MySqlData. -->
+    <!-- System.Drawing.Common is tranistive dependency required by OpenTelemetry.Instrumentation.StackExchangeRedis. -->
```
Comment (pjanotti) [inline src/CommonExcludedAssets.props:14]: `System.Drawing.Common`? 🤔

**Review by pjanotti — APPROVED** [review]
👍

**Review by zacharycmontoya — APPROVED** [review]
LGTM, this is a great solution

File: src/CommonExcludedAssets.props (hunk 1 above)
Comment (Kielek) [reply src/CommonExcludedAssets.props:14]: Yes... Deeply hidden. I hope I have didnt make any issue while copy pasting chain.

`OpenTelemetry.Instrumentation.StackExchangeRedis/1.0.0-rc9.10`
-> `StackExchange.Redis/2.1.58`
-> `System.Diagnostics.PerformanceCounter/4.7.0`
-> `System.Configuration.ConfigurationManager/4.7.0`
-> `System.Security.Permissions/4.7.0`
-> `System.Windows.Extensions/4.7.0`
->  `System.Drawing.Common/4.7.0`

**Pull request merged**